### PR TITLE
Allow non-root user to run git clone plugin

### PR DESCRIPTION
**Pull request by shubham149**

### CODE REVIEW

**Comment by tonglil** [comment]
Hi @bradrydzewski, just an FYI that this was a breaking change for anyone that set the `workspace` to `/root` or did cloning operations to `/root` ($HOME).

Context: this was necessary when passing along a file (ie credentials) that was sourced automatically at $HOME/..., but since the default is `/drone/src` and `$HOME` wasn't `/drone` until this PR, we need to now change configs and other references from `/root` to `/drone`.

**Comment by tonglil** [comment]
Another scenario where this causes an issue:

- Don't change the workspace
- Clone
- Use `node` image to run `npm ci`
- `npm ci` fails because the node image's user is `root` and wants to install dependency cache at `$HOME/.npm` which is `/root/.npm`.

Workaround is to configure the `.npmrc` or `mkdir -p /root`, but I can see this happening elsewhere too.
Not sure if setting workspace to `/root` can solve the issue?

FWIW the node image creates the user but doesn't specify it as the default: https://github.com/nodejs/docker-node/blob/8b68fca7a5089bd8795ae85b55617314e966487b/16/buster/Dockerfile

**Comment by tonglil** [comment]
Hi @bradrydzewski, just an FYI that this caused some issues compared to prior behavior?

**Comment by tonglil** [comment]
1. I think changing the user and $HOME messed up the use of .netrc and .ssh-key files?
```
npm ERR! Host key verification failed.
npm ERR! fatal: Could not read from remote repository.
npm ERR! 
npm ERR! Please make sure you have the correct access rights
```

**Comment by bradrydzewski** [comment]
Hi Tony, I just ran a test and I am able to use the netrc file to clone a private repository in my pipeline.  I used the below yaml file for testing, where _bradrydzewski/petstore-app_ is private.

```
kind: pipeline
type: docker
name: default

steps:
- name: clone2
  image: golang
  commands:
  - git clone https://github.com/bradrydzewski/petstore-app.git
  - ls -la petstore-app
  - ls -la /root
```

These were the results:

Screen Shot 2021-05-07 at 3 54 26 PM

The change we made in this pull request writes the .netrc to the home directory of an unprivileged user, however, this only impacts this plugin. The reason is that every pipeline step is responsible for reading the netrc environment variable and writing to its home directory [1] (this is what allows the netrc to be read by images regardless of USER defined in the dockerfile). In terms of ssh keys, Drone only supports the netrc file.

I would be interested to further understand the issue you are facing.  In the meantime, you can pin your runners to use an older version of the git plugin, such as `drone/git:1.2.0`, by using [DRONE_RUNNER_CLONE_IMAGE](https://docs.drone.io/runner/docker/configuration/reference/drone-runner-clone-image/).

[1] https://github.com/drone-runners/drone-runner-docker/blob/master/engine/compiler/shell/shell.go#L38

**Comment by tonglil** [comment]
Thanks, I'm still debugging and trying to elaborate on the details of the issues we're seeing to provide you with reproducible examples.

In the mean time we will use the tag.

**Comment by tonglil** [comment]
Here is a scenarios that fails now compared to before.
I cannot tell what the issue is.

Here is a 4 pipelines to illustrate the issues (file level perms and git dialing):

```yaml
---
kind: pipeline
type: docker
name: old-clone-latest-node

clone:
  disable: true

steps:
  - name: clone
    image: drone/git:1.2.0-linux-amd64
    pull: always
    commands:
      - /usr/local/bin/clone

  - name: step_2
    # image: node:10.18.0
    image: node:latest
    pull: always
    commands:
      - npm install

---
kind: pipeline
type: docker
name: old-clone-old-node

clone:
  disable: true

steps:
  - name: clone
    image: drone/git:1.2.0-linux-amd64
    pull: always
    commands:
      - /usr/local/bin/clone

  - name: step_2
    image: node:10.18.0
    pull: always
    commands:
      - npm install

---
kind: pipeline
type: docker
name: new-clone-latest-node

clone:
  disable: true

steps:
  - name: clone
    image: drone/git:latest
    pull: always
    commands:
      - /usr/local/bin/clone

  - name: step_2
    image: node:latest
    pull: always
    commands:
      - npm install

---
kind: pipeline
type: docker
name: new-clone-old-node

clone:
  disable: true

steps:
  - name: clone
    image: drone/git:latest
    pull: always
    commands:
      - /usr/local/bin/clone

  - name: step_2
    image: node:10.18.0
    pull: always
    commands:
      - npm install
```

The repo only has .drone.yml and package.json:
```
{
  "name": "test",
  "version": "0.1.12",
  "description": "x",
  "main": "index.js",
  "license": "MIT",
  "dependencies": {
    "text": "github:requirejs/text"
  }
}
```

FWIW the node image creates the user but doesn't specify it as the default: https://github.com/nodejs/docker-node/blob/8b68fca7a5089bd8795ae85b55617314e966487b/16/buster/Dockerfile

File: posix/clone [hunk 1]
```diff
@@ -1,15 +1,19 @@
...
 
-# if the home directory is not set (which should
-# never be the case) we default to /root
+# we default home directory to /home/drone
 
-if [[ -z "${HOME}" ]]; then
-	echo "HOME directory not set; default to /root"
-	export HOME=/root
+if [ "$HOME" != "/home/drone" ]; then
+	export HOME=/home/drone
```
Comment (tonglil) [inline posix/clone:16]: AFAIK `/home/drone` isn't persisted between steps? Does it make sense to set it or create it in the docker file either?

**Comment by tonglil** [comment]
Where we notice the .netrc isn't being used any more is when trying to install node js dependencies from private GitHub repos.

For example, with a package.json:
```
{
  "name": "test",
  "version": "0.1.12",
  "description": "x",
  "main": "index.js",
  "license": "MIT",
  "dependencies": {
    "private-package": "https://github.com/my-org/private-repo.git"
  }
}
```

The error when running `npm install` is:

```
npm ERR! Error while executing:
npm ERR! /usr/bin/git ls-remote -h -t https://github.com/my-org/private-repo.git
npm ERR! 
npm ERR! remote: Invalid username or password.
npm ERR! fatal: Authentication failed for 'https://github.com/my-org/private-repo.git/'
npm ERR! 
npm ERR! exited with error code: 128
```

When previously it worked successfully.

File: posix/clone (hunk 1 above)
Comment (bradrydzewski) [reply posix/clone:16]: Only the /drone workspace is persisted between steps.  The plugin previously wrote to the /root which was not persisted between step either.

**Comment by bradrydzewski** [comment]
hey Tony, I took a look at your first example and see it uses the git+ssh clone url.  The netrc file is only used by git+https and would therefore never be used when cloning a git+ssh repository url.  I would expect this operation to always fail without an ssh key, which is not something Drone natively supports.

```text
npm ERR! Error while executing:
npm ERR! /usr/bin/git ls-remote -h -t ssh://git@github.com/my-org/private-repo.git
```

Is there something happening under the hood that I'm not seeing?

**Comment by tonglil** [comment]
My understanding of npm is that the `github:org/repo` syntax will iterate through different ways of authenticating, and I'm guessing when it fails https it tries ssh.

The 4 pipelines demonstrate the change in behavior: https://github.com/drone/drone-git/pull/39#issuecomment-834752047

**Comment by tonglil** [comment]
I created a sample on cloud.drone.io for you to view: https://cloud.drone.io/tonglil/drone-git-debug/1
Repo: https://github.com/tonglil/drone-git-debug
It uses the basic package.json from the npm docs: https://docs.npmjs.com/cli/v7/configuring-npm/package-json#github-urls

**Comment by bradrydzewski** [comment]
hey Tony, keep in mind that Drone Cloud is for public repositories only.  The netrc file is _only_ made available to a pipeline if the repository is private.  If you can reproduce a problem on drone cloud, it would imply this has nothing to do with netrc or authentication.

I ran a test with a private repository and I was able to reproduce the problem with `node:latest` as it appears to always try and use the git+ssh url even though I explicitly configured the git+https url in my package.json.  Note in the logs that I am able to git-ls remote into the repository just fine in the same step, which indicates the netrc is working just fine.

Screen Shot 2021-05-07 at 6 26 27 PM

Screen Shot 2021-05-07 at 6 29 19 PM

Screen Shot 2021-05-07 at 6 28 22 PM

I then changed the image in my pipeline to node:14 and everything worked as expected:

Screen Shot 2021-05-07 at 6 18 12 PM

**Comment by bradrydzewski** [comment]
I used debug mode to ssh into a live pipeline that failed with `node:latest`.  I deleted the netrc (just to rule out authentication) and then I re-ran `npm install` which failed again

```
root@8881046f6b6c:/drone# rm /root/.netrc
root@8881046f6b6c:/drone# npm install
The authenticity of host 'github.com (140.82.112.4)' can't be established.
RSA key fingerprint is SHA256:nThbg6kXUpJWGl7E1IGOCspRomTxdCARLviKw6E5SY8.
[..................] - idealTree:drone: sill idealTree buildDeps
npm ERR! code 128
npm ERR! command failed
npm ERR! command git ls-remote ssh://git@github.com/requirejs/text.git
npm ERR! Host key verification failed.
npm ERR! fatal: Could not read from remote repository.
npm ERR! 
npm ERR! Please make sure you have the correct access rights
npm ERR! and the repository exists.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-05-07T22_55_48_132Z-debug.log
```

I then was able to `cat` the log file and I can see that npm is attempting to use `git+ssh` address to connect to the repository instead of `git+https`.  I guess the question I have is, why are older versions of npm using `git+https` and newer versions using `git+ssh`? 

```
root@8881046f6b6c:/drone# cat /root/.npm/_logs/2021-05-07T22_55_48_132Z-debug.log
0 verbose cli [ '/usr/local/bin/node', '/usr/local/bin/npm', 'install' ]
1 info using npm@7.7.6
2 info using node@v15.14.0
3 timing npm:load:whichnode Completed in 2ms
4 timing config:load:defaults Completed in 2ms
5 timing config:load:file:/usr/local/lib/node_modules/npm/npmrc Completed in 1ms
6 timing config:load:builtin Completed in 1ms
7 timing config:load:cli Completed in 2ms
8 timing config:load:env Completed in 0ms
9 timing config:load:file:/drone/.npmrc Completed in 1ms
10 timing config:load:project Completed in 2ms
11 timing config:load:file:/root/.npmrc Completed in 0ms
12 timing config:load:user Completed in 0ms
13 timing config:load:file:/usr/local/etc/npmrc Completed in 1ms
14 timing config:load:global Completed in 1ms
15 timing config:load:validate Completed in 0ms
16 timing config:load:setEnvs Completed in 1ms
17 timing config:load Completed in 9ms
18 timing npm:load:configload Completed in 9ms
19 timing npm:load:setTitle Completed in 0ms
20 timing npm:load:setupLog Completed in 1ms
21 timing npm:load:cleanupLog Completed in 4ms
22 timing npm:load:configScope Completed in 0ms
23 timing npm:load:projectScope Completed in 1ms
24 timing npm:load Completed in 18ms
25 timing config:load:flatten Completed in 5ms
26 timing arborist:ctor Completed in 1ms
27 timing idealTree:init Completed in 28ms
28 timing idealTree:userRequests Completed in 0ms
29 silly idealTree buildDeps
30 silly fetch manifest text@requirejs/text
31 silly placeDep ROOT text@ OK for: my_package@1.0.0 want: requirejs/text
32 timing idealTree:#root Completed in 28464ms
33 timing idealTree:node_modules/text Completed in 0ms
34 timing idealTree:buildDeps Completed in 28466ms
35 timing idealTree:fixDepFlags Completed in 1ms
36 timing idealTree Completed in 28496ms
37 timing command:install Completed in 28534ms
38 verbose stack Error: command failed
38 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/node_modules/@npmcli/promise-spawn/index.js:64:27)
38 verbose stack     at ChildProcess.emit (node:events:369:20)
38 verbose stack     at maybeClose (node:internal/child_process:1067:16)
38 verbose stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:301:5)
39 verbose cwd /drone
40 verbose Linux 4.19.76-linuxkit
41 verbose argv "/usr/local/bin/node" "/usr/local/bin/npm" "install"
42 verbose node v15.14.0
43 verbose npm  v7.7.6
44 error code 128
45 error command failed
46 error command git ls-remote ssh://git@github.com/requirejs/text.git
47 error Host key verification failed.
```

**Comment by bradrydzewski** [comment]
Is this an npm issue? https://github.com/npm/cli/issues/2610

**Comment by tonglil** [comment]
It could be related? But the failing version is npm: '6.13.4'.

A bunch of different symptoms cropped up as our runners pulled the new plugins/git image and I'm investigating them one at a time, trying to make sense of what's changed.

In this thread of js/npm, we have a repo that uses private repo dependencies and it also has a package-lock.json file. When the package-lock.json exists, `npm install` with the new git plugin fails with node:13.8.0 and node:14 but not the old git plugin. When the package-lock.json is removed, `npm install` works on all versions with all git plugins.

A lot to try to debug right now so we're just going to revert to 1.2.0 as a workaround until it's more clear what the issue is.

**Comment by bradrydzewski** [comment]
I was reading through the issue and they mention that they have the same problem with npm6 if a developer is using npm7 and commits the updated package-lock.json to the repository

> While I do use npm v7 locally and it forms package-lock.json, code builds may run other versions. I just checked one AWS CodeBuild we have and it runs node v12.18.0 (with npm 6.14.4). It is tripped by the existing package-lock.json and works only with npm i, which is not what I want. https://github.com/npm/cli/issues/2610#issuecomment-797072195

**Comment by tonglil** [comment]
Thanks for taking a look, I appreciate you taking the time to address our concerns.
At this time there's no actionable change to make.
I would be curious to see if anyone else crops up to report any issues.
I passed on this information to the team that's out this week and they will follow up.
I will be out next week.

**Comment by bradrydzewski** [comment]
Thanks for helping debug!  Just to document in this thread, the reason we made these changes is because previous versions of the plugin did not work on openshift due to restrictions against running in root, which impacted a segment of our users.

If this ends up being related to the npm issue, we may recommend relying on the 1.0.0 version of the docker image (via global runner configuration) until npm releases a fix.  If we find out this is unrelated to npm or that npm wontfix, we may have to publish two versions of the clone image (one that is root and another that is unprivileged)

**Comment by akhenakh** [comment]
Thanks @tonglil thanks @bradrydzewski 

Can you confirm the correct version to pin for now: 
`DRONE_RUNNER_CLONE_IMAGE= drone/git:1.2.0-linux-amd64`

**Comment by bradrydzewski** [comment]
I think you need to ping to `DRONE_RUNNER_CLONE_IMAGE=drone/git:1.0.0-linux-amd64`.  The 1.2.0 image used alpine 3.13 as a base and it had a DNS bug that prevented github.com from resolving.

**Comment by akhenakh** [comment]
You mean 1.0.0 (1.1.0 does not exist)?

https://hub.docker.com/layers/drone/git/1.0.0-linux-amd64/images/sha256-44603a87a8d0ce477e25a43362df7dd8b04d1c3ae00435c3d9abdd48b400df91?context=explore

**Comment by bradrydzewski** [comment]
@flyingfishflash this image is meant for internal use only and is not something you should be using directly in your pipeline steps section. You should instead be using something like [`alpine/git`](https://hub.docker.com/r/alpine/git).  See [this example](https://docs.drone.io/pipeline/docker/syntax/cloning/#custom-logic) in our documentation.

**Comment by flyingfishflash** [comment]
@bradrydzewski thanks for the reply, I'll switch out of the internal image. I probably used it to avoid pulling another image.

**Pull request merged**